Tutorial - Python analytics (#7873)
PostHog/posthog.com: contents/tutorials/python-analytics.md

---

> title: How to set up analytics in Python and Flask

Author: ```suggestion
title: How to set up Python analytics in Flask
```

```bash
pip3 install Flask
```

Create a new folder for your project called `flask-analytics` and set up a basic project structure:

```bash
mkdir flask-analytics
cd flask-analytics
mkdir templates static
touch app.py
cd ./templates
touch login.html
touch dashboard.html
cd ..
```

Then, we set up our app routes in `app.py`:

```python file=app.y
from flask import Flask, render_template, request, redirect, session, url_for

app = Flask(__name__)
app.secret_key = 'a_super_secret_key!' # you don't need to replace this

@app.route('/')
def index():
    return render_template('login.html')

@app.route('/login', methods=['POST'])
def login():
    session['name'] = request.form['name']
    session['email'] = request.form['email']
    session['company_name'] = request.form['company_name']
    return redirect(url_for('dashboard'))

@app.route('/dashboard')
def dashboard():
    if 'name' in session:
        return render_template('dashboard.html', session=session)
    return redirect(url_for('index'))

@app.route('/api/dashboard', methods=['POST'])
def api_dashboard():
    # Placeholder for API logic. We'll add this later
    return '', 204

if __name__ == '__main__':

>     app.run(debug=True)

Reviewer: Apparently Mac uses port 5000 for its [control centre](https://stackoverflow.com/questions/72795799/how-to-solve-403-error-with-flask-in-python), so adjusting this to make it work
```suggestion
    app.run(port=8000, debug=True)
```

        <input type="text" name="company_name" required>

        <button type="submit">Log in</button>
    </form>
</body>
</html>
```

```html file=templates/dashboard.html
<!DOCTYPE html>
<html>
<head>
    <title>Dashboard</title>
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.6.0/jquery.min.js"></script>
</head>
<body>
    <h1>Welcome, {{ session['name'] }} from {{ session['company_name'] }}!</h1>

    <form id="dashboardForm" action="{{ url_for('api_dashboard') }}" method="POST">
        <input type="hidden" name="email" value="{{ session['email'] }}">
        <input type="hidden" name="name" value="{{ session['name'] }}">
        <input type="hidden" name="company_name" value="{{ session['company_name'] }}">
        <button type="submit">Click Me</button>
    </form>

    <script>
    $(document).ready(function() {
        $('#dashboardForm').on('submit', function(e) {
            e.preventDefault();
            $.ajax({
                url: $(this).attr('action'),
                type: 'POST',
                data: $(this).serialize(),
            });
        });
    });
    </script>
</body>
</html>
```

> Run `python3 app.py` and navigate to `http://localhost:5000` to see our app in action. Enter anything on the login page to save some session details.

Reviewer: ```suggestion
Run `python3 app.py` and navigate to `http://localhost:8000` to see our app in action. Enter anything on the login page to save some session details.
```

```py file=apo.py
package main
from flask import Flask, render_template, request, redirect, session, url_for
from posthog import Posthog

posthog = Posthog(
  '<ph_project_api_key>', 
  host='<ph_instance_address>'
)

# rest of your code
```

With this set up, we're ready to capture events.

## 3. Implement the event capture code

To show how to capture events with PostHog, we capture an event when the button on the `dashboard` page is clicked. To do this, we call [`posthog.capture()`](/docs/libraries/python#capturing-events):

```python file=app.py
@app.route('/api/dashboard', methods=['POST'])
def api_dashboard():
    email = request.form.get('email')
    posthog.capture(
        email, 
        'home_api_called'
    )
    return '', 204
```

With this set up, refresh your app and click the button on the `dashboard` page a few times. You should now see the captured event in your [PostHog activity tab](https://us.posthog.com/events).

<ProductScreenshot
  imageLight={EventsInPostHogLight} 
  imageDark={EventsInPostHogDark} 
  alt="Events in PostHog" 
  classes="rounded"
/>

> > 💡 **PostHog tip: Setting the correct `DistinctId`** 

Reviewer: ```suggestion
> 💡 **PostHog tip: Setting the correct `distinct_id`** 
```

> 

> > When calling `posthog.capture()`, you need to provide a `distinctId` argument. This is a unique identifier for your user and enables you to correctly attribute events to them. 

Reviewer: ```suggestion
> When calling `posthog.capture()`, you need to provide a `distinct_id` argument. This is a unique identifier for your user and enables you to correctly attribute events to them. 
```

As an example, we add the user's name as an event property:

```python file=app.py
@app.route('/api/dashboard', methods=['POST'])
def api_dashboard():
    email = request.form.get('email')
    name = request.form.get('name')
    posthog.capture(
        email, 
        'home_api_called',
        properties={
            "user_name": name
        }
    )
    return '', 204
```

### Capturing group events

[Groups](/docs/product-analytics/group-analytics) are a powerful feature in PostHog that aggregate events based on entities, such as organizations or companies. This is especially helpful for B2B SaaS apps, where often you want to view insights such as `number of active companies` or `company churn rate`.

To enable group analytics, you'll need to [upgrade](https://us.posthog.com/organization/billing) your PostHog account to include them. This requires entering your credit card, but don't worry, we have a [generous free tier](/pricing) of 1 million events per month – so you won't be charged anything yet.

> To create groups in PostHog, simply include them in your code when capturing events by setting the `$groups` argument:

Reviewer: ```suggestion
To create groups in PostHog, simply include them in your code when capturing events by setting the `groups` argument:
```

Author: needs to have the $
